File: sdk-python/arcflow/provider.py
"""LLM provider configuration (Sprint 6). Credentials from environment only."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Union

from arcflow.constants import (
    PROVIDER_DEFAULT_MAX_TOKENS,
    PROVIDER_DEFAULT_TEMPERATURE,
)
from arcflow.exceptions import ProviderConfigurationError


@dataclass(frozen=True)
class OpenAI:
    """OpenAI provider; reads ``OPENAI_API_KEY`` from the environment."""

    model: str
    max_tokens: int = PROVIDER_DEFAULT_MAX_TOKENS
    temperature: float = PROVIDER_DEFAULT_TEMPERATURE

    def __post_init__(self) -> None:
        if not self.model.strip():
            raise ProviderConfigurationError(
                "[ArcFlow] OpenAI model must be a non-empty string. "
                "Example: OpenAI(model='gpt-4o')."
            )
        if not 0.0 <= self.temperature <= 1.0:
            raise ProviderConfigurationError(
                f"[ArcFlow] OpenAI temperature must be between 0.0 and 1.0. "
                f"Got {self.temperature}."
            )
        if self.max_tokens < 1:
            raise ProviderConfigurationError(
                f"[ArcFlow] OpenAI max_tokens must be at least 1. Got {self.max_tokens}."
            )

    def binding_tuple(self) -> tuple[str, str, int, float]:
        return ("openai", self.model, self.max_tokens, self.temperature)


@dataclass(frozen=True)
class Anthropic:
    """Anthropic provider; reads ``ANTHROPIC_API_KEY`` from the environment."""

    model: str
    max_tokens: int = PROVIDER_DEFAULT_MAX_TOKENS
    temperature: float = PROVIDER_DEFAULT_TEMPERATURE

    def __post_init__(self) -> None:
        if not self.model.strip():
            raise ProviderConfigurationError(
                "[ArcFlow] Anthropic model must be a non-empty string. "
                "Example: Anthropic(model='claude-3-5-sonnet-20241022')."
            )
        if not 0.0 <= self.temperature <= 1.0:
            raise ProviderConfigurationError(
                f"[ArcFlow] Anthropic temperature must be between 0.0 and 1.0. "
                f"Got {self.temperature}."
            )
        if self.max_tokens < 1:
            raise ProviderConfigurationError(
                f"[ArcFlow] Anthropic max_tokens must be at least 1. Got {self.max_tokens}."
            )

    def binding_tuple(self) -> tuple[str, str, int, float]:
        return ("anthropic", self.model, self.max_tokens, self.temperature)


@dataclass(frozen=True)
class Gemini:
    """Gemini provider; reads ``GEMINI_API_KEY`` from the environment."""

    model: str
    max_tokens: int = PROVIDER_DEFAULT_MAX_TOKENS
    temperature: float = PROVIDER_DEFAULT_TEMPERATURE

    def __post_init__(self) -> None:
        if not self.model.strip():
            raise ProviderConfigurationError(
                "[ArcFlow] Gemini model must be a non-empty string. "
                "Example: Gemini(model='gemini-1.5-pro')."
            )
        if not 0.0 <= self.temperature <= 1.0:
            raise ProviderConfigurationError(
                f"[ArcFlow] Gemini temperature must be between 0.0 and 1.0. "
                f"Got {self.temperature}."
            )
        if self.max_tokens < 1:
            raise ProviderConfigurationError(
                f"[ArcFlow] Gemini max_tokens must be at least 1. Got {self.max_tokens}."
            )

    def binding_tuple(self) -> tuple[str, str, int, float]:
        return ("gemini", self.model, self.max_tokens, self.temperature)


ProviderConfig = Union[OpenAI, Anthropic, Gemini]

File: sdk-typescript/arcflow/provider.ts
import {
  PROVIDER_DEFAULT_MAX_TOKENS,
  PROVIDER_DEFAULT_TEMPERATURE,
} from "./constants.js";
import { ProviderConfigurationError } from "./exceptions.js";

export interface ProviderConfig {
  model: string;
  maxTokens?: number;
  temperature?: number;
}

function validateProvider(kind: string, config: ProviderConfig) {
  if (!config.model.trim()) {
    throw new ProviderConfigurationError(
      `[ArcFlow] ${kind} model must be a non-empty string.`,
    );
  }
  const temperature = config.temperature ?? PROVIDER_DEFAULT_TEMPERATURE;
  if (temperature < 0 || temperature > 1) {
    throw new ProviderConfigurationError(
      `[ArcFlow] ${kind} temperature must be between 0.0 and 1.0. Got ${temperature}.`,
    );
  }
  const maxTokens = config.maxTokens ?? PROVIDER_DEFAULT_MAX_TOKENS;
  if (maxTokens < 1) {
    throw new ProviderConfigurationError(
      `[ArcFlow] ${kind} maxTokens must be at least 1. Got ${maxTokens}.`,
    );
  }
  return { maxTokens, temperature };
}

export class OpenAI {
  readonly model: string;
  readonly maxTokens: number;
  readonly temperature: number;

  constructor(config: ProviderConfig) {
    const validated = validateProvider("OpenAI", config);
    this.model = config.model.trim();
    this.maxTokens = validated.maxTokens;
    this.temperature = validated.temperature;
  }

  bindingRow() {
    return {
      kind: "openai",
      model: this.model,
      maxTokens: this.maxTokens,
      temperature: this.temperature,
    };
  }
}

export class Anthropic {
  readonly model: string;
  readonly maxTokens: number;
  readonly temperature: number;

  constructor(config: ProviderConfig) {
    const validated = validateProvider("Anthropic", config);
    this.model = config.model.trim();
    this.maxTokens = validated.maxTokens;
    this.temperature = validated.temperature;
  }

  bindingRow() {
    return {
      kind: "anthropic",
      model: this.model,
      maxTokens: this.maxTokens,
      temperature: this.temperature,
    };
  }
}

export class Gemini {
  readonly model: string;
  readonly maxTokens: number;
  readonly temperature: number;

  constructor(config: ProviderConfig) {
    const validated = validateProvider("Gemini", config);
    this.model = config.model.trim();
    this.maxTokens = validated.maxTokens;
    this.temperature = validated.temperature;
  }

  bindingRow() {
    return {
      kind: "gemini",
      model: this.model,
      maxTokens: this.maxTokens,
      temperature: this.temperature,
    };
  }
}

export type Provider = OpenAI | Anthropic | Gemini;
